deepspeed/runtime/engine.py

```diff
@@ -1757,6 +1757,26 @@ def is_gradient_accumulation_boundary(self):
             return self._is_gradient_accumulation_boundary
 
     def set_gradient_accumulation_boundary(self, is_boundary):
+        """Manually overrides the DeepSpeed engine's gradient accumulation boundary state, this is an optional
+        feature and should be used with care. The state should be set before to the intended
+        value before each forward/backward. The final fordward/backward should have the
+        boundary state set to True. See example below:
+
+        .. code-block:: python
+
+        engine.set_gradient_accumulation_boundary(False)
+        for _ in range(gradient_accumulation_steps - 1):
+            micro_batch = next(data_loader)
+            loss = engine(micro_batch)
+            engine.backward(loss)
+        engine.set_gradient_accumulation_boundary(True)
+        micro_batch = next(data_loader)
+        loss = engine(micro_batch)
+        engine.backward(loss)
```

Comment:
For completeness, should this snippet not include `engine.step()`?

Reply:
Oh good call, that's an important bit, in this case we only need to call `engine.step()` once at the end.

Reply:
Also, would the following be correct usage of this api?
```python
   for i, micro_batch in enumerate(next(data_loader)):
        is_gradient_accumulation_boundary = (i+1 % gradient_accumulation_steps) == 0
        engine.set_gradient_accumulation_boundary(is_gradient_accumumation_boundary)
        loss = engine(micro_batch)
        engine.backward(loss)
        engine.step()

```

Reply:
Yep, that would work too. In the case of megatron and fairseq they both call their optimizer step in completely different code paths then the main fwd/bwd passes so it's a bit more aligned with the style in the docstring (without a final engine.step) I guess.